src/plugins/discord/sqliteMirror.js

```diff
@@ -0,0 +1,491 @@
+            nick
+          FROM members
+          WHERE user_id = :user_id
+          `
+      )
+      .get({user_id: userId});
+
+    if (member === undefined) {
+      return undefined;
+    } else {
+      return {
+        user: NullUtil.get(this.user(userId), `No user_id found for ${userId}`),
+        nick: member.nick,
+      };
+    }
+  }
+
+  channels(): $ReadOnlyArray<Model.Channel> {
+    return this._db
+      .prepare(
+        dedent`\
+          SELECT
+            id,
+            type,
+            name
+          FROM channels`
+      )
+      .all();
+  }
+
+  messages(channel: Model.Snowflake): $ReadOnlyArray<Model.Message> {
+    return this._db
+      .prepare(
+        dedent`\
+          SELECT
+            id,
+            channel_id,
+            author_id,
+            non_user_author,
+            timestamp_ms,
+            content
+          FROM messages
+          WHERE channel_id = :channel_id`
+      )
+      .all({channel_id: channel})
+      .map((m) => ({
+        id: m.id,
+        channelId: m.channel_id,
+        authorId: m.author_id,
+        nonUserAuthor: m.non_user_author === 1,
+        timestampMs: m.timestamp_ms,
+        content: m.content,
+        reactionEmoji: this.reactionEmoji(m.channel_id, m.id),
+        mentions: this.mentions(m.channel_id, m.id),
+      }));
+  }
+
+  reactions(
+    channel: Model.Snowflake,
+    message: Model.Snowflake
+  ): $ReadOnlyArray<Model.Reaction> {
+    return this._db
+      .prepare(
+        dedent`\
+          SELECT
+            channel_id,
+            message_id,
+            author_id,
+            emoji
+          FROM message_reactions
+          WHERE channel_id = :channel_id
+            AND message_id = :message_id`
+      )
+      .all({channel_id: channel, message_id: message})
+      .map((r) => ({
+        channelId: r.channel_id,
+        messageId: r.message_id,
+        authorId: r.author_id,
+        emoji: Model.refToEmoji(r.emoji),
+      }));
+  }
+
+  mentions(
+    channel: Model.Snowflake,
+    message: Model.Snowflake
+  ): $ReadOnlyArray<Model.Snowflake> {
+    return this._db
+      .prepare(
+        dedent`\
+          SELECT user_id
+          FROM message_mentions
+          WHERE channel_id = :channel_id
+            AND message_id = :message_id`
+      )
+      .all({channel_id: channel, message_id: message})
+      .map((res) => res.user_id);
+  }
+
+  addUser(user: Model.User) {
+    this._db
+      .prepare(
+        dedent`\
+          REPLACE INTO users (
+              id,
+              username,
+              discriminator,
+              bot
+          ) VALUES (
+              :id,
+              :username,
+              :discriminator,
+              :bot
+          )
+        `
+      )
+      .run({
+        id: user.id,
+        username: user.username,
+        discriminator: user.discriminator,
+        bot: Number(user.bot),
+      });
+  }
+
+  addMember(member: Model.GuildMember) {
```

Comment:
Because `Model.GuildMember` contains the full User object and the membership, suggest calling `addUser` under the hood, rather than require the caller does this.

Reply:
Yes, good call. Updated.